A C++ style guide for Toit (#1150)
toitlang/toit: src/README.md

# Toit C++ style guide

When contributing code to Toit, please follow the style of the
existing C++ code as much as possible.  This makes code easier
to read and update for other project members.

Here we mention some of the style rules that we try to follow.
In general it is more important to follow the
style of the code already in Toit than to adhere to the style
guide.

## Language features

For performance reasons, we compile with `-fno-exceptions` and so C++ exception
handling is not available.

We also compile with `-fno-rtti`, though this might change at some point.

We do not allow use of the C++ standard library on the device, for reasons of

> code size.  In particular this means no use of `std::string`.

Reviewer: Not just. `std::string` uses `malloc` which is a more important reason not to use it.

```
int bar() {
  foo(56, 612);  // bad
  foo(42, 103);  // This is a sentence.
  return 89;
}
```

For line comments (`//`) leave at least two spaces between the code and the comment.
If there are not alignment issues, leave exactly two spaces.  After the double
slash, leave one space.

```
void baz() {
  my_function(56, 612);//Too cramped.
  foo(42, 103);  // That's more like it.
  f(87);         // Can leave more space for alignment.
}
```

## Naming

Classes have an initial capital letter and camel case.  Abbreviations are treated as normal words with an initial capital letter.

```
class foo;             // No.
class Bar;             // Yes.
class FooBar;          // Yes, camel.
class HttpConnection;  // Yes, capitalize HTTP like it was a regular word.
```

Methods and functions are written with lower case and underscores.

```
class Foo {
 public:
  int method_a();  // Yes.
  int MethodB();   // No, use lower case.
  int methodC();   // No, use underscores.
};
```

Local variables and arguments can be short, but not too short.  Avoid needless abbreviations.
For example, `address` is usually better than `addr`.  Single-letter loop variables like `i`

> and `j` are OK for very small loops, but it probably makes sense to give a better name for
> loops of 10 lines or more.

Reviewer: It should be ok to use `i`, if the first thing I do is `auto x = foo[i]`.
```suggestion
and `j` are OK if their uses are in proximity to the `for` keyword.
```

Top-level constants are named with `ALL_CAPS_AND_UNDERSCORES`.

## Readability

Try to avoid unnamed bool arguments.

```
  my_function(42, true);    // What does this mean?

  bool force = true;

>   my_function(42, force);  // More readable.

Reviewer: ?
```suggestion
  my_function(42, force);  // More readable.

  bool force;
  my_function(42, force=false);  
```

```

## Toit-specific

### Compiler

The Toit compiler is not run on the memory-constrained devices, so it may
use standard C++ library features, like std::string.  But don't go mad
(no Boost, please).

The Toit compiler is designed to run for a short time and restart frequently.
Therefore it does not normally free memory.  This simplifies C++ a lot.
Thus there is no need for 'smart pointers'.

### VM/OS

On the device memory is very restricted.  Therefore any `malloc` may fail
at any time, except perhaps at system startup.  The code should cope with
a failed `malloc` by triggering a GC.

Do not use plain `new`, but instead use `_new`.  If you are integrating
third party code that uses `new` you will hit a fatal assertion.  This
is because such code generally cannot cope with memory exhaustion without
crashing.

Some uses of C++ closures will cause the compiler to generate
calls to `new` that will crash on memory shortage.  Avoid capturing
too many variables to avoid this.  Otherwise you will hit an assert.

> Always check the return value of `_new` for a null pointer.  In this
> case the constructor has not been called.

Reviewer: ```suggestion
Always check the return value of `_new` for a null pointer.  In this
case the constructor has not been called. This is not required in the compiler, which can assume
that there is infinite memory.
```
